In store pickup intersection search graphql

swnsma opened the pull request:
### Description (*)
Add support of intersection assignment filter for GraphQl endpoint.
Add functional tests coverage.

Request for GraphQl schema changes: https://github.com/magento/architecture/pull/419


### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/inventory#3169: In store pickup intersection search graphql

Commit:
In-Store-Pickup.
Add GraphQl support for products assignment intersection.

Commit:
Fix tests.

sidolov commented:
@magento run all tests

Commit:
Fix static tests.

swnsma commented on InventoryInStorePickupGraphQl/Model/Resolver/PickupLocations/SearchRequest/ProductsInfo.php:
```diff
@@ -0,0 +1,81 @@
+    private function getExtensionAttributes(
+        SearchRequestBuilderInterface $searchRequestBuilder
+    ): SearchRequestExtensionInterface {
+        $builderCopy = clone $searchRequestBuilder;
```
There is a problem that we can override extension attributes which were previously added to the builder.
I see 3 possible solutions for the problem:
1) Current implementation - build copy of search request and get extension attributes from it.
2) Ignore the problem. Could cause a problem for extension developers and solution providers - only one class will be allowed to set extension attributes.
3) Rethink service contracts. Probably, add new interface which will have 'getExtensionAttributes' or 'getData' for the builder.
Any suggestions?

novikor replied:
I think we must follow option 3, but still have no idea what the best solution is.

swnsma commented:
@magento run all tests

novikor reviewed (commented):
Looks fine for now.

swnsma commented:
@magento create ticket

Commit:
Refactoring.
Incapsulate internal logic in single class and expose extension point for extension attributes.

swnsma committed:
Merge branch '1.2-develop' into in-store-pickup-intersection-search-graphql

Commit:
Refactoring.
Change paramenters list of SearchRequestBuilder::resolve.
Revert test fuxtes.

Commit:
Merge branch 'in-store-pickup-intersection-search-graphql' of github.com:magento/inventory into in-store-pickup-intersection-search-graphql

Commit:
Refactoring.
Rename method.

novikor commented:
@magento create issue

swnsma commented:
@magento run all tests

Commit:
Fix static tests.

novikor commented:
@magento run all tests

novikor commented on InventoryInStorePickupGraphQl/Model/Resolver/PickupLocations/SearchRequestBuilder.php:
```diff
@@ -0,0 +1,257 @@
+     *
+     * @return SearchRequestBuilderInterface
+     */
+    public function getBuilderFromArgument(string $fieldName, array $argument): SearchRequestBuilderInterface
```
`SearchRequestBuilder` with a single public method that returns another `SearchRequestBuilder`?
Let's rename it into something less confusing.

novikor commented on InventoryInStorePickupGraphQl/Model/Resolver/PickupLocations/SearchRequestBuilder.php:
```diff
@@ -0,0 +1,257 @@
+     *
+     * @return void
+     */
+    private function resolveAttributeValue(
```
Please see my general CR comment regarding this method.

novikor reviewed (changes requested):
I like your idea about the `SearchRequest` refactoring, but I see some points to improve.
I suggest to do the following:
- Replace `\Magento\InventoryInStorePickupGraphQl\Model\Resolver\PickupLocations\SearchRequestBuilder::resolveAttributeValue` with a chain of responsibility (in the fact, just restore 6 classes we had before)
- Make the chain not extendable to prevent using it instead of the `ExtensionProvider`. I mean something like that:
```php
    private const RESOLVERS_MAPPING = [
        'RADIUS_FIELD' => RADIUS_FIELD_RESOLVER::class,
        'SEARCH_TERM' => SEARCH_TERM_RESOLVER::class,
        'SORT_FIELD' => SORT_FIELD_RESOLVER::class,
        'AREA_FIELD' => AREA_FIELD_RESOLVER::class,
        'FILTERS_FIELD' => FILTERS_FIELD_RESOLVER::class,
        'CURRENT_PAGE' => CURRENT_PAGE_RESOLVER::class,
        'PAGE_SIZE' =>PAGE_SIZE_RESOLVER::class
    ];
```

swnsma committed:
Merge branch '1.2-develop' into in-store-pickup-intersection-search-graphql

Commit:
magento/inventory#3169: In store pickup intersection search graphql.
Apply code review changes.

Commit:
Merge branch 'in-store-pickup-intersection-search-graphql' of github.com:magento/inventory into in-store-pickup-intersection-search-graphql

swnsma commented:
@magento run all tests

swnsma commented:
@magento run all tests

novikor commented on InventoryInStorePickupGraphQl/Model/Resolver/PickupLocations.php:
```diff
@@ -47,27 +41,25 @@ class PickupLocations implements \Magento\Framework\GraphQl\Query\ResolverInterf
     ) {
-        $this->searchRequestResolver = $searchRequestResolver;
-        $this->searchRequestBuilder = $searchRequestBuilder;
+        $this->searchRequestBuilderResolver = $searchRequestBuilderResolver;
```
What about `$searchRequestBuilder` instead of `$searchRequestBuilderResolver`?

novikor reviewed (changes requested):
Thank you, looks fine. All we need now are green tests.

novikor commented:
@magento run Functional Tests B2B

novikor commented:
@magento run Functional Tests CE

novikor commented:
@magento run Functional Tests B2B

slavvka commented:
@magento give me test instance

slavvka commented:
@magento give me new test instance

swnsma committed:
Merge branch '1.2-develop' into in-store-pickup-intersection-search-graphql

swnsma committed:
Merge branch '1.2-develop' into in-store-pickup-intersection-search-graphql

Commit:
Merge branch '1.2-develop' into in-store-pickup-intersection-search-graphql

sidolov commented:
@magento run all tests

novikor commented:
@magento run Functional Tests EE

novikor commented:
@magento run Functional Tests B2B

swnsma commented:
@magento run Functional Tests EE

swnsma commented:
@magento run Functional Tests EE
